blender: allow functional declaration within withPackages

the-furry-hubofeverything opened the pull request:
## Description of changes
**Originally included in #257780, separated since it used to introduce breaking changes.**
(not anymore, now `blender-with-packages` is being deprecated.)

This allows the following syntax to be used
```
blender.withPackages(ps: [ ps.bpycv ])
``` 
instead of  
```
blender-with-packages { packages = [ python310Packages.bpycv ]; }
```

Allowing `withPackages` to use the pythonPackages from the `blender` package.

I've updated `passthru.tests.render` for bpycv to use the new implementation.

blender-with-packages is deprecated, as the new implementation breaks the currently used syntax, and it would have trouble integrating with #257780, where multiple versioned variants of blender would be involved.

Also passed through pythonPackages to pass thorugh any overrides applied to blender's current `python310Packages`, and without splicing issues (discussed in #257780)


*to test, enter `nix-build -I nixpkgs=./ -E 'with (import <nixpkgs> {}); blender.withPackages(ps: [ps.bpycv])'`, then opening `blender-wrapped`, go to the "scripting" workplace (top tabs), and try importing `bpycv` through the python terminal` 

or try `nix-build -A blender.pythonPackages.bpycv.tests.render`

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

the-furry-hubofeverything committed:
blender: allow functional declaration within withPackages

Based on #257780, separated since it introduces significant changes.

bpycv: update passthru.tests.render

blender-with-packages: deprecated
it is still backwards compatible, but no longer preferred.

the-furry-hubofeverything commented:
While rebasing the other PR onto this I noticed I forgot to remove `blender-with-packages` from bpycv

the-furry-hubofeverything commented:
That should be everything, mostly the breaking changes part.

the-furry-hubofeverything commented:
Failing checks may be related to https://github.com/NixOS/nixpkgs/pull/264290, rebasing...

lucasew commented:
Result of `nixpkgs-review pr 264292` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>

lucasew commented:
bpycv test is working fine

pbsds commented on pkgs/applications/misc/blender/default.nix:
```diff
@@ -18,10 +18,12 @@
 }:
 
 let
   python = python310Packages.python;
```
```suggestion
  pythonPackages = python310Packages;
  inherit (pythonPackages) python;
```

pbsds commented on pkgs/applications/misc/blender/default.nix:
```diff
@@ -191,6 +193,9 @@ stdenv.mkDerivation (finalAttrs: rec {
   passthru = {
     inherit python;
```
```suggestion
    inherit python pythonPackages;
```

pbsds reviewed (commented):
LGTM, mainly nits from me

the-furry-hubofeverything commented:
@pbsds nits picked

Followed @K900's advice and made `blender.withPackages` backwards compatible with `blender-with-packages`. All functionality is preserved... I think.

This would no longer be a breaking change, but `blender-with-packages` will be deprecated.

the-furry-hubofeverything commented:
Oh dear

the-furry-hubofeverything commented:
Should be alright now

Additionally - 
- Moved the manual entry into "Backward Incompatibilities" instead of "Other notable changes"
- Moved blender-with-packages from `all-packages.nix` to `aliases.nix`

Mindavi commented on pkgs/top-level/aliases.nix:
```diff
@@ -92,6 +92,7 @@ mapAliases ({
   bird2 = bird; # Added 2022-02-21
   bitwig-studio1 = throw "bitwig-studio1 has been removed, you can upgrade to 'bitwig-studio'"; # Added 2023-01-03
   bitwig-studio2 = throw "bitwig-studio2 has been removed, you can upgrade to 'bitwig-studio'"; # Added 2023-01-03
+  blender-with-packages = args: lib.warn "blender-with-packages is deprecated and will be removed in later versions" blender.withPackages (_: args.packages) { name = args.name; }; # Added 2023-10-30
```
Consider advertising the alternative here too: withPackages. Then it's clearer that there's a new method instead of it sounding like it's not supported at all anymore.

the-furry-hubofeverything replied:
Changed message to be more clearer, thought that line has become quite a bit longer...

K900 replied:
Don't worry about it. Also, the message should probably refer people to `blender.withPackages`.

the-furry-hubofeverything replied:
> Don't worry about it. Also, the message should probably refer people to `blender.withPackages`.

The selected line was outdated. 

Current version is -

`blender-with-packages is deprecated in favor of blender.withPackages, and will be removed in future releases`

K900 replied:
Maybe something like `blender-with-packages is deprecated, please use blender.withPackages(ps: [ ps.example ])`? Something that has an example, so it's clearer for how to migrate.

the-furry-hubofeverything replied:
> Maybe something like `blender-with-packages is deprecated, please use blender.withPackages(ps: [ ps.example ])`? Something that has an example, so it's clearer for how to migrate.

I would think that people would look at the changelog for it... hmm.

I want to add "Similar to python3.withPackages" to the end of it too, would that be too long?

the-furry-hubofeverything replied:
I just think it's should be `blender-with-packages is deprecated, please use blender.withPackages, which behaves similar to python3.withPackages.`, probably removing the last part for brevity

the-furry-hubofeverything commented:
@K900 @Mindavi 

Sorry about the quick changes, but since you folks made some points to clarify the documentation, I've updated both the documentation and aliases.nix to clarify the new function better.

K900 commented on pkgs/applications/misc/blender/wrapper.nix:
```diff
@@ -2,17 +2,15 @@
-, packages ? []
+, extraModules ? []
 }:
+{ name ? "wrapped" }:
```
Is this actually useful? I feel like moving it to `overrideAttrs` is fine, if people really really want this behavior for some reason.

the-furry-hubofeverything replied:
I have no idea. @lucasew wrote that line, and I may need a sec to figure out how to migrate this to work with overrideAttrs. 

I don't know a way to override `withPackages` directly, but it probably will require overriding blender's `buildEnv` that's declared in `default.nix`, so it'd be something like `(blender.override { wrapperName = "withBpycv" }).withPackages ( ... )` or something

If we remove that outright it will be a breaking change

the-furry-hubofeverything replied:
@K900 I don't know how github comments work, do people preferred being pinged? Should I ping you?

K900 replied:
You get emails when people reply to your thread, so no need for a ping. You can basically achieve this (I think?) with `blender.overrideAttrs(old: { name = "blender-fancy-${old.version}"; })`, which can be done in the `blender-with-packages` shim.

lucasew replied:
> Is this actually useful? I feel like moving it to `overrideAttrs` is fine, if people really really want this behavior for some reason.

It's for the suffix of some stuff like desktop items so when installing it differentiate between each wrapper generated. As far as I remember.

the-furry-hubofeverything replied:
I'm going to sweep everything under `pname`

the-furry-hubofeverything replied:
> > Is this actually useful? I feel like moving it to `overrideAttrs` is fine, if people really really want this behavior for some reason.
> 
> It's for the suffix of some stuff like desktop items so when installing it differentiate between each wrapper generated. As far as I remember.

Wait a minute... could this mean that multiple version of blender could theoretically be installed with this wrapper?

This... could change things

(I should probably stay within scope for now though)

the-furry-hubofeverything replied:
> You get emails when people reply to your thread, so no need for a ping. You can basically achieve this (I think?) with `blender.overrideAttrs(old: { name = "blender-fancy-${old.version}"; })`, which can be done in the `blender-with-packages` shim.

now its `(blender.withPackages ( ... )).overrideAttrs ({ pname = "blender-whatever"; })`

the-furry-hubofeverything commented:
Alright, no more `name`, if people really wants to they could rename the whole thing by using `overrideAttrs`

the-furry-hubofeverything commented:
Here's the diff: https://github.com/NixOS/nixpkgs/compare/b60862aeeb5ddb9b10a536f7eb99e283f289b3a8..327dbdf62c271a95d681a120b7a28d9c0b85c3fd

the-furry-hubofeverything commented:
I think a refactor of wrapper.nix is required, but I think that should be done later, not with this PR.

K900 commented on pkgs/top-level/aliases.nix:
```diff
@@ -92,6 +92,7 @@ mapAliases ({
   bird2 = bird; # Added 2022-02-21
   bitwig-studio1 = throw "bitwig-studio1 has been removed, you can upgrade to 'bitwig-studio'"; # Added 2023-01-03
   bitwig-studio2 = throw "bitwig-studio2 has been removed, you can upgrade to 'bitwig-studio'"; # Added 2023-01-03
+  blender-with-packages = args: lib.warn "blender-with-packages is deprecated in favor of blender.withPackages, which behaves similarly to python3.withPackages" (blender.withPackages (_: args.packages)).overrideAttrs (lib.optionalAttrs (args ? name) { pname = "blender-" + args.name; }); # Added 2023-10-30
```
Maybe add some line breaks here? This line really feels like a lot.

the-furry-hubofeverything replied:
> Don't worry about it. Also, the message should probably refer people to blender.withPackages.
told ya. I'll do something about it

the-furry-hubofeverything replied:
Does this work?

https://github.com/NixOS/nixpkgs/blob/521cd5a7d283a5ef655eacba58dd488d980a03d4/pkgs/top-level/aliases.nix#L95-L97

K900 replied:
Definitely more readable, yeah. Also, I'd probably drop the `python3.withPackages` bit entirely and just provide an example - something like "...deprecated in favor of `blender.withPackages`, e.g. `blender.withPackages(ps: [ ps.honk ])`"

the-furry-hubofeverything replied:
> Definitely more readable, yeah. Also, I'd probably drop the `python3.withPackages` bit entirely and just provide an example - something like "...deprecated in favor of `blender.withPackages`, e.g. `blender.withPackages(ps: [ ps.honk ])`"

I just feel iffy about referring to a specific package for a general example

K900 replied:
I don't really see a problem with that personally. If you really wanted to be fancy about it, you could generate the expression from the passed-in package names, I guess (but I feel like this would be trying way too hard lmao)

the-furry-hubofeverything replied:
Ah, I know what to do!

the-furry-hubofeverything replied:
https://github.com/NixOS/nixpkgs/blob/fe4c776ae9c4289d848096fba1621b3c5c6d0bde/pkgs/top-level/aliases.nix#L96

How's that?
